source/simulator/assemblers/advection.cc

```diff
@@ -468,6 +468,307 @@ namespace aspect
+      const FEValuesExtractors::Scalar solution_field = advection_field.scalar_extractor(introspection);
+      MaterialModel::MeltOutputs<dim> *melt_outputs = scratch.material_model_outputs.template get_additional_output<MaterialModel::MeltOutputs<dim>>();
+
+      for (unsigned int q=0; q<n_q_points; ++q)
+        {
+          // precompute the values of shape functions and their gradients.
+          // We only need to look up values of shape functions if they
+          // belong to 'our' component. They are zero otherwise anyway.
+          // Note that we later only look at the values that we do set here.
+          for (unsigned int i=0, i_advection=0; i_advection<advection_dofs_per_cell;/*increment at end of loop*/)
+            {
+              if (fe.system_to_component_index(i).first == solution_component)
+                {
+                  if (use_supg)
+                    scratch.laplacian_phi_field[i_advection] = trace(scratch.finite_element_values[solution_field].hessian (i,q));
+
+                  scratch.grad_phi_field[i_advection] = scratch.finite_element_values[solution_field].gradient (i,q);
+                  scratch.phi_field[i_advection]      = scratch.finite_element_values[solution_field].value (i,q);
+                  ++i_advection;
+                }
+              ++i;
+            }
+
+          const double density_c_P              =
+            ((advection_field_is_temperature)
+             ?
+             scratch.material_model_outputs.densities[q] *
+             scratch.material_model_outputs.specific_heat[q]
+             :
+             1.0);
+
+          Assert (density_c_P >= 0,
+                  ExcMessage ("The product of density and c_P needs to be a "
+                              "non-negative quantity."));
+
+          const double reaction_term =
+            ((advection_field_is_temperature)
+             ?
+             0.0
+             :
+             scratch.material_model_outputs.reaction_terms[q][advection_field.compositional_variable]);
+
+          const double field_term_for_rhs
+            = (use_bdf2_scheme ?
+               (scratch.old_field_values[q] *
+                (1 + time_step/old_time_step)
+                -
+                scratch.old_old_field_values[q] *
+                (time_step * time_step) /
+                (old_time_step * (time_step + old_time_step)))
+               :
+               scratch.old_field_values[q]);
+
+
+          const double rho_s = scratch.material_model_outputs.densities[q];
+          const double rho_f = melt_outputs->fluid_densities[q];
+          const Tensor<1,dim> gravity = this->get_gravity_model().gravity_vector (scratch.finite_element_values.quadrature_point(q));
+          const double K_D = melt_outputs->permeabilities[q] / melt_outputs->fluid_viscosities[q];
+          Tensor<1,dim> current_u = scratch.current_velocity_values[q];
+          Tensor<1,dim> current_u_f = current_u - K_D * (rho_s * gravity - rho_f * gravity);
```

Comment:
The Darcy equation we looked at earlier has the gradient of pressure instead of`rho_s * g`. Is using `rho_s * g` equivalent to assuming all pressure variations are hydrostatic? If so, should we consider adding in an option to use the full pressure gradient? As currently constructed, this formulation would not have dynamic pressure variations included, but perhaps we don't need them as `current_u ` will include these effects to some degree? 

@jdannberg - thoughts on this?